### ISSUE TITLE
When doing 'real' async in function with a HttpClient.SendAscync the logger and IServiceCollection are disposed before the function ends

### ISSUE THREAD

**Issue by RikShoeby** [Azure/azure-functions-dotnet-worker#600]
When doing 'real' async in function with a HttpClient.SendAscync the logger and IServiceCollection are disposed before the function ends.

I have some HttpClients:
`private readonly Dictionary<string, HttpClient> _httpClients;`

When calling the SendAsync with await the function does not work including the logging
`var response = await _httpClients[urlName].SendAsync(request);`

When calling the SendAsync without await the function does work including the logging, but now the azure function is not really async (what we want ofcourse)
`var response = _httpClients[urlName].SendAsync(request).Result;`

In dotnet core 3.1 this did work

**Comment by RikShoeby** [comment]
@kshyju Of-course I can give more information. 
Add the moment we also have big problems with the retry deliveries. I made a small repro for that:
https://github.com/Azure/azure-functions-dotnet-worker/issues/607

The logging from the ILogger did not show up in the logstream in the function app and not in the app insights.
For the IServiceCollection I did get an exceptions that the ServiceCollection was disposed. For we inject the correct mapper instead of search for it in a generic way, what is oke because the azure functions only use one or two mappers.
![image](https://user-images.githubusercontent.com/89585999/131317854-4cb9547d-98f0-41cc-a6c0-72c422baeee3.png)

I will take a look into the above and see if that's helping, after that I will try to make a small reproduction sample.

But why are you use the 'using around the http client? This is a bad practice because you want to reuse the client..?
Note that I use functions with a servicebus trigger.

**Comment by kshyju** [comment]
The `IHttpClientFactory` implementation efficiently handles the lifetime of `HttpMessageHandler`. So it is okay to dispose the httpclient you create using this factory. In my example, it will also dispose the `HttpResponseMessage` instance. 

Please see the [official doc](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests) for detailed explanation about the benefits of using IHttpClientFactory.

> For the IServiceCollection I did get an exceptions that the ServiceCollection was disposed.

Could you please share a minimal repro (a github repro link would be great) where you reproduce this problem ?

**Comment by RikShoeby** [comment]
Thanks for the information, I will take a look into that and enable the mapper code again for the IServiceCollection. Then I will look into reproducing it again.

I have found a programming bug. We used async Task/ValueTask evereywere  (api's, services, controllers etc) except the azure functions. This seems to fix most of our problems. 

Again I will take a look to the use of the IHttpClientFactory instead of our own create a client one's per baseUrl endpoint. And how to refactor that..
